examples/bmi-for-nextgen.ipynb

```diff
@@ -0,0 +1,1696 @@
+    "\n",
+    "In `snowPy` all variables had a global scope and a limited lifetime (only while the script was executing). For `snowBMI`, we converted the model to a package with a snow class populated by the model state variables we need access to. \n",
+    "\n",
+    "```python\n",
+    "class Snow(object):\n",
+    "    \"\"\"Snow model class.\"\"\"\n",
+    "\n",
+    "    def __init__(\n",
+    "        self, rs_method=1, rs_thresh=2.5, snow_thresh_max=1.5, rain_thresh_min=4.5,\n",
+    "            ddf_max=1, ddf_min=0, tair_melt_thresh=1, swe_init=0, dayofyear=274, year=2016,\n",
+    "    ):\n",
+    "\n",
+    "        self._rs_method = rs_method\n",
+    "        self._rs_thresh = rs_thresh\n",
+    "        self._snow_thresh_max = snow_thresh_max\n",
+    "        self._rain_thresh_min = rain_thresh_min\n",
+    "        self._ddf_max = ddf_max\n",
+    "        self._ddf_min = ddf_min\n",
+    "        self._tair_melt_thresh = tair_melt_thresh\n",
+    "\n",
+    "        self._time = 0.0\n",
+    "        self._time_step = 86400\n",
+    "        self._dayofyear = dayofyear\n",
+    "        self._year = year\n",
+    "\n",
+    "        self._tair_c = np.zeros(1, dtype=float)\n",
+    "        self._ppt_mm = np.zeros(1, dtype=float)\n",
+    "        self._swe_mm = np.zeros(1, dtype=float)\n",
+    "        swe_tmp = np.zeros(1, dtype=float)\n",
+    "        swe_tmp[0, ] = swe_init\n",
+    "        self._swe_mm = swe_tmp\n",
+    "        self._melt_mm = np.zeros(1, dtype=float)\n",
+    "```\n",
+    "\n",
+    "This means that once the `snowBMI` model has been initialized, we can access the data using BMI functions, which we'll explore now.\n",
+    "\n",
+    "## 3.3 Implementing BMI functions\n",
+    "\n",
+    "As of this writing, there are 41 different functions in the BMI version 2.0 specification. Obviously we don't have time to cover all of them, so we'll cover some of the important ones for NextGen. \n",
+    "\n",
+    "NextGen does many things, but here are some highlights along with select (i.e., not all) relevant BMI functions:\n",
+    "\n",
+    "1. It initializes models\n",
+    "  - BMI::initialize()\n",
+    "2. It controls model runtime\n",
+    "  - BMI::update()\n",
+    "3. It closes models when they're completed to free up memory and other resources\n",
+    "  - BMI::finalize()\n",
+    "4. It passes forcing data to models (and can pass data between coupled models in a stack)\n",
+    "  - BMI::get_value(), BMI::set_value()\n",
+    "5. It can perform automated unit conversion when passing data\n",
+    "  - BMI::get_var_units()\n",
+    "6. It keeps the clock running for all models in an instance\n",
+    "  - BMI::get_current_time(), BMI::get_time_step()\n",
+    "\n",
+    "We will discuss these further in the sections below. A few notes/reminders first:\n",
+    "\n",
+    "- *You must implement all BMI functions even if they only return BMI_FAILURE* for BMI to work.\n",
```

Comment:
Would it be worth pointing out the BMI Minimal subclass I've worked on which "implements" many of the functions either with generic implementations or raising errors for others (like grid/mesh functions?). I've only built this for Python at the moment, but may extend this into some of the other supported languages at some point.  This might also over complicate things?  But dev tools are always useful early in the game IMHO

Reply:
Yes! At the very least, I can share the link to BMI Minimal. Is it on your GitHub?